### PR TITLE
Refactor of LakeFS cmd to allow implementing #2364 - Dumping configuration on startup and errors

### PR DESCRIPTION
**Pull request by edmondo1984**
This PR was created after trying to address issue #2364 . When trying to review the code, I notice it got to a complexity where it was hard to modify, unit testing and debugging. The rootCmd function was 180 lines to set-up a complex graph of objects, many of them would require 6+ parameters to be constructed.

Therefore, before adding dumping configuration, it could make sense to improve the "shape of the code". This PR adopts this approach: 
- An application package would contain "higher level services" and factory methods to create those services 
- A cmd package containing the code that invokes the application package, getting significantly simplified

The "application" name is stolen from domain driven design methodology, which suggests that the "application layer" is where you compose "domain" services to create functionalities that are exposed to the outside world.  

As a side node, this PR also address the duplication in setting up new users in `setup.go` and `superuser.go`

### CODE REVIEW

File: cmd/lakefs/application/lakefsserver.go
Comment (itaiad200) [inline cmd/lakefs/application/lakefsserver.go:9]:
```suggestion
func NewLakeFSHTTPServer(listenAddress string, s3GatewayDomainNames []string, s3gatewayHandler http.Handler, apiHandler http.Handler) *http.Server {
```

File: cmd/lakefs/application/auth.go
Comment (itaiad200) [inline cmd/lakefs/application/auth.go:11]: We organize the imports differently. A block for core libraries and others. See the `run.go` file in this PR for reference.

File: cmd/lakefs/cmd/run.go
Comment (itaiad200) [inline cmd/lakefs/cmd/run.go:37]: It's not the `RootCmd`. Can you revert this function change?

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks for making this change. Please see my comments, we shouldn't panic on all database errors, some are even expected. 
Also please fix the failing tests. I think we have a problem with running the Nessie workflow for forked PRs, so once the go tests pass, I'll try to run the tests for that workflow myself

File: cmd/lakefs/cmd/run.go
Comment (edmondo1984) [reply cmd/lakefs/cmd/run.go:37]: Ops, yes. If you have nothing in comment I will leave the function pattern. It's easy to to test, I will just rename it as NewRunCmd()

File: cmd/lakefs/cmd/run.go
Comment (edmondo1984) [reply cmd/lakefs/cmd/run.go:56]: Could you please help me to understand how is the flow supposed to continue when the database connectivity is not available?

Comment (edmondo1984) [reply cmd/lakefs/cmd/run.go:56]: Thanks. In terms of the general approach, are you ok with it?

Comment (itaiad200) [reply cmd/lakefs/cmd/run.go:56]: I'm fine with the approach you took for the refactoring. I think we may need 3 PRs here though. 
- Refactoring which will also include other commands that use the same building blocks like setup, superuser, etc..
- Handling DB errors correctly.
- Dumping the configuration with secrets omitted.

Comment (edmondo1984) [reply cmd/lakefs/cmd/run.go:56]: 👍  let me work on it

Comment (edmondo1984) [reply cmd/lakefs/cmd/run.go:56]: @itaiad200 would it be ok if we keep this PR for the first two item in the list, and I will open a separate PR for the configuration dumping?

Comment (itaiad200) [reply cmd/lakefs/cmd/run.go:56]: Sure thing. Thanks for taking care of that!
Please change the title and description of this PR.

**Comment by itaiad200** [comment]
@edmondo1984 some tests are still failing. Can you take a look at the failing checks and fix them?
Any PR must complete all checks before it can be merged.

File: cmd/lakefs/application/lakefscmdcontext.go
Comment (itai-david) [inline cmd/lakefs/application/lakefscmdcontext.go:10]:
```suggestion
type LakeFSCmdContext struct {
```

File: cmd/lakefs/application/lakefscmdcontext.go
Comment (itai-david) [inline cmd/lakefs/application/lakefscmdcontext.go:16]:
```suggestion
func NewLakeFSCmdContext(ctx context.Context, cfg *config.Config, logger logging.Logger) LakeFsCmdContext {
```

File: cmd/lakefs/application/lakefscmdcontext.go
Comment (itai-david) [inline cmd/lakefs/application/lakefscmdcontext.go:17]:
```suggestion
	return LakeFSCmdContext{
```

File: cmd/lakefs/application/lakefscmdcontext.go
Comment (itai-david) [inline cmd/lakefs/application/lakefscmdcontext.go:11]: I believe [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#contexts) suggest contexts should not be added as a struct member

Comment (edmondo1984) [reply cmd/lakefs/application/lakefscmdcontext.go:11]: This is a great feedback, thanks. Did you use a tool to find out?

**Review by itai-david — COMMENTED** [review]
Thank you, @edmondo1984 for initiating this PR
Refactoring are (almost) always welcome and this one is defintely required here, and thank you for implementing it
Please see some minor comments/suggestions inline.
Apart from these, as I was going over the extracted code, I got to think that at least some of it is general enough and maybe can be extracted even further, to outside of the lakefs/cmd tree, to pkg, where it might be even furtherly reused.
Did you consider this option?

File: cmd/lakefs/application/lakefscmdcontext.go
Comment (itai-david) [reply cmd/lakefs/application/lakefscmdcontext.go:11]: No tool was used. This one came up in a conversation with @nopcoder, which suggested this standard

**Comment by edmondo1984** [comment]
> Thank you, @edmondo1984 for initiating this PR Refactoring are (almost) always welcome and this one is defintely required here, and thank you for implementing it Please see some minor comments/suggestions inline. Apart from these, as I was going over the extracted code, I got to think that at least some of it is general enough and maybe can be extracted even further, to outside of the lakefs/cmd tree, to pkg, where it might be even furtherly reused. Did you consider this option?

No I didn't, do you think it's simple enough to be added to the current PR or should we create a separate issue for that? I am always afraid of "trying to do too much in a single PR"

**Comment by itai-david** [comment]
> No I didn't, do you think it's simple enough to be added to the current PR or should we create a separate issue for that? I am always afraid of "trying to do too much in a single PR". Can you also maybe give me some more details about this further improvements? I didn't explore the lakefs/pkg folder much

Nothing concrete at the moment, but the first thing comes to my mind is some testing facilities that can utilize various abilities you refactored, such as DB connection creation or Block Adapter creation. The later is actually created by testing code, for testing purposes, but in a much lighter use.
And I definitely agree with the smaller PR approach and this can definitely be resolved as a different issue with further refactoring suggestion

File: cmd/lakefs/cmd/run.go
Comment (edmondo1984) [inline cmd/lakefs/cmd/run.go:112]: Was there a missing os.Exit(1) here? I have added to the code @itai-david

**Comment by itaiad200** [comment]
Created https://github.com/treeverse/lakeFS/pull/3148 for the tests to run

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks @edmondo1984 !
I ran your tests in #3148 and they seem to be passing! See few comments that I made.
Since this is quite a big PR that refactors a lot of our critical parts of the system, I would request another set of eyes to review this.

File: cmd/lakefs/cmd/import.go
Comment (edmondo1984) [reply cmd/lakefs/cmd/import.go:104]: Are we always exiting after a logger.Fatal?

File: cmd/lakefs/cmd/users.go
Comment (edmondo1984) [reply cmd/lakefs/cmd/users.go:32]: Golint would cause a failure "do not trigger dynamic error", I had to specifically to add a new error. Let me try to commit your suggestion and see what happens, maybe my code was slightly different

File: cmd/lakefs/cmd/run.go
Comment (edmondo1984) [reply cmd/lakefs/cmd/run.go:58]: The logging is done in the function, because it was duplicated I refactored it. If I extract it I will re-duplicate it (run and import). Are you ok with that?

File: cmd/lakefs/cmd/import.go
Comment (itaiad200) [reply cmd/lakefs/cmd/import.go:104]: `logger.Fatal` exists with 1 `statusCode`

File: cmd/lakefs/cmd/users.go
Comment (edmondo1984) [reply cmd/lakefs/cmd/users.go:48]: They seem not, at least  in master:

https://github.com/treeverse/lakeFS/blob/master/cmd/lakefs/cmd/superuser.go#L39

**Review by nopcoder — CHANGES_REQUESTED** [review]
First, thanks - a lot of effort was put into this one. Comments inside the code.
The PR is big and it looks like it tries to address a lot of things, I suggest we will do by listing the steps and create a pr for each one or review each change.
- Package structure 
- Service initialization and composition
- Testable cobra command

File: cmd/lakefs/cmd/import.go
Comment (itai-david) [inline cmd/lakefs/cmd/import.go:138]:
```suggestion
	prefixes, err := getPrefixes(prefixFile)
```

File: cmd/lakefs/cmd/import.go
Comment (itai-david) [inline cmd/lakefs/cmd/import.go:137]: We can delete this one and use the short-declare in the next line instead

**Review by itai-david — COMMENTED** [review]
Hi @edmondo1984, and thanks again for putting the effort into this refactor
I added some comments, most of which are not crucial, but I think they can improve the final result a bit
Thanks again

**Comment by itaiad200** [comment]
Closing due to lack of activity. It can be reopen if necessary.

**Pull request closed without merging**
